crates/apollo-compiler/src/validation/field.rs

```diff
@@ -210,3 +214,122 @@ pub fn validate_field_definitions(
+            hir::TypeDefinition::EnumTypeDefinition(_) if !leaf => {
+                Some("This field is an enum and cannot select any fields")
+            }
+            hir::TypeDefinition::ScalarTypeDefinition(_) if !leaf => {
+                Some("This field is a scalar and cannot select any fields")
+            }
+            hir::TypeDefinition::ObjectTypeDefinition(_) if leaf => {
+                Some("This field is an object and must select fields")
+            }
+            hir::TypeDefinition::InterfaceTypeDefinition(_) if leaf => {
+                Some("This field is an interface and must select fields")
+            }
+            hir::TypeDefinition::UnionTypeDefinition(_) if leaf => {
+                Some("This field is an union and must select fields")
+            }
+            _ => None,
+        };
+        if let Some(error) = leaf_validation_error {
+            let name = field.name.src.clone();
+            let diagnostic_data = if leaf {
+                DiagnosticData::MandatorySubselection {
+                    name,
+                    ty: type_name.clone(),
+                }
+            } else {
+                DiagnosticData::NoSubselectionAllowed {
+                    name,
+                    ty: type_name.clone(),
+                }
+            };
+            let diagnostic = ApolloDiagnostic::new(db, field.loc.into(), diagnostic_data)
+                .label(Label::new(field.loc, error));
+            let diagnostic = if let Some(type_def_loc) = type_def.loc() {
+                diagnostic.label(Label::new(
+                    type_def_loc,
+                    format!("`{}` declared here", &type_name),
+                ))
+            } else {
+                diagnostic
+            };
+            return Some(diagnostic);
+        }
+    }
+    None
+}
+
+#[cfg(test)]
+mod test {
+    use crate::ApolloCompiler;
+
+    #[test]
+    fn it_validates_enum_as_leaf_field() {
+        let input = r#"
+query SelectionOfEnum {
+  pet {
+    name
+  }
+}
+
+type Query {
+  pet: Pet,
+}
+
+enum Pet {
+    CAT
+    DOG
+    FOX
+}
+"#;
+
+        let mut compiler = ApolloCompiler::new();
+        compiler.add_document(input, "schema.graphql");
+
+        let diagnostics = compiler.validate();
+        for diagnostic in &diagnostics {
+            println!("{diagnostic}");
+        }
+
+        assert_eq!(diagnostics.len(), 1)
+    }
+
+    #[test]
+    fn it_validates_union_as_leaf_field() {
+        let input = r#"
+query SelectionOfEnum {
+  animal
+}
+
+type Query {
+  animal: CatOrDog,
+}
+
+type Cat { id: String! }
+type Dog { id: String! }
+union CatOrDog = Cat | Dog
+"#;
+
+        let mut compiler = ApolloCompiler::new();
+        compiler.add_document(input, "schema.graphql");
+
+        let diagnostics = compiler.validate();
+        for diagnostic in &diagnostics {
+            println!("{diagnostic}");
+        }
+
+        assert_eq!(diagnostics.len(), 1)
+    }
+}
```

Comment:
these can live in `test_data` dir

Reply:
They already are. I put some tests here for a faster feedback during development like
```
cargo watch -x "test it_validates_enum_as_leaf_field --package apollo-compiler -- --nocapture"
```

I can remove them now.

Reply:
https://github.com/apollographql/apollo-rs/pull/465/commits/1114eca6719d9aca3e9e1f70d4e4095a4c86e5a8